File: java/client/src/main/java/glide/api/models/commands/ZAddOptions.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.api.models.commands;

import glide.api.commands.SortedSetBaseCommands;
import glide.api.models.GlideString;
import glide.utils.ArgsBuilder;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import lombok.Builder;
import lombok.RequiredArgsConstructor;

/**
 * Optional arguments to {@link SortedSetBaseCommands#zadd(String, Map, ZAddOptions, boolean)},
 * {@link SortedSetBaseCommands#zadd(String, Map, ZAddOptions)} and {@link
 * SortedSetBaseCommands#zaddIncr(String, String, double, ZAddOptions)}
 *
 * @see <a href="https://valkey.io/commands/zadd/">valkey.io</a>
 */
@Builder
public final class ZAddOptions {
    /**
     * Defines conditions for updating or adding elements with {@link SortedSetBaseCommands#zadd}
     * command.
     */
    private final ConditionalChange conditionalChange;

    /**
     * Specifies conditions for updating scores with zadd {@link SortedSetBaseCommands#zadd} command.
     */
    private final UpdateOptions updateOptions;

    @RequiredArgsConstructor
    public enum ConditionalChange {
        /**
         * Only update elements that already exist. Don't add new elements. Equivalent to <code>XX
         * </code> in the Valkey API.
         */
        ONLY_IF_EXISTS("XX"),
        /**
         * Only add new elements. Don't update already existing elements. Equivalent to <code>NX</code>
         * in the Valkey API.
         */
        ONLY_IF_DOES_NOT_EXIST("NX");

        private final String valkeyApi;
    }

    @RequiredArgsConstructor
    public enum UpdateOptions {
        /**
         * Only update existing elements if the new score is less than the current score. Equivalent to
         * <code>LT</code> in the Valkey API.
         */
        SCORE_LESS_THAN_CURRENT("LT"),
        /**
         * Only update existing elements if the new score is greater than the current score. Equivalent
         * to <code>GT</code> in the Valkey API.
         */
        SCORE_GREATER_THAN_CURRENT("GT");

        private final String valkeyApi;
    }

    /**
     * Converts ZaddOptions into a String[].
     *
     * @return String[]
     */
    public String[] toArgs() {
        if (conditionalChange == ConditionalChange.ONLY_IF_DOES_NOT_EXIST && updateOptions != null) {
            throw new IllegalArgumentException(
                    "The GT, LT, and NX options are mutually exclusive. Cannot choose both "
                            + updateOptions.valkeyApi
                            + " and NX.");
        }

        List<String> optionArgs = new ArrayList<>();

        if (conditionalChange != null) {
            optionArgs.add(conditionalChange.valkeyApi);
        }

        if (updateOptions != null) {
            optionArgs.add(updateOptions.valkeyApi);
        }

        return optionArgs.toArray(new String[0]);
    }

    /**
     * Converts ZaddOptions into a GlideString[].
     *
     * @return GlideString[]
     */
    public GlideString[] toArgsBinary() {
        return new ArgsBuilder().add(toArgs()).toArray();
    }
}

File: go/api/options/zadd_options.go
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

package options

import (
	"errors"

	"github.com/valkey-io/valkey-glide/go/utils"
)

// Optional arguments to `ZAdd` in [SortedSetCommands]
type ZAddOptions struct {
	conditionalChange ConditionalChange
	updateOptions     UpdateOptions
	changed           bool
	incr              bool
	increment         float64
	member            string
}

func NewZAddOptionsBuilder() *ZAddOptions {
	return &ZAddOptions{}
}

// `conditionalChange` defines conditions for updating or adding elements with `ZADD` command.
func (options *ZAddOptions) SetConditionalChange(c ConditionalChange) *ZAddOptions {
	options.conditionalChange = c
	return options
}

// `updateOptions` specifies conditions for updating scores with zadd command.
func (options *ZAddOptions) SetUpdateOptions(u UpdateOptions) *ZAddOptions {
	options.updateOptions = u
	return options
}

// `Changed` changes the return value from the number of new elements added to the total number of elements changed.
func (options *ZAddOptions) SetChanged(ch bool) (*ZAddOptions, error) {
	if options.incr {
		return nil, errors.New("changed cannot be set when incr is true")
	}
	options.changed = ch
	return options, nil
}

// `INCR` sets the increment value to use when incr is true.
func (options *ZAddOptions) SetIncr(incr bool, increment float64, member string) (*ZAddOptions, error) {
	if options.changed {
		return nil, errors.New("incr cannot be set when changed is true")
	}
	options.incr = incr
	options.increment = increment
	options.member = member
	return options, nil
}

// `ToArgs` converts the options to a list of arguments.
func (opts *ZAddOptions) ToArgs() ([]string, error) {
	args := []string{}
	var err error

	if opts.conditionalChange == OnlyIfExists || opts.conditionalChange == OnlyIfDoesNotExist {
		args = append(args, string(opts.conditionalChange))
	}

	if opts.updateOptions == ScoreGreaterThanCurrent || opts.updateOptions == ScoreLessThanCurrent {
		args = append(args, string(opts.updateOptions))
	}

	if opts.changed {
		args = append(args, ChangedKeyword)
	}

	if opts.incr {
		args = append(args, IncrKeyword, utils.FloatToString(opts.increment), opts.member)
	}

	return args, err
}

// A ConditionalSet defines whether a new value should be set or not.
type ConditionalChange string

const (
	// Only update elements that already exist. Don't add new elements. Equivalent to "XX" in the Valkey API.
	OnlyIfExists ConditionalChange = "XX"
	// Only add new elements. Don't update already existing elements. Equivalent to "NX" in the Valkey API.
	OnlyIfDoesNotExist ConditionalChange = "NX"
)

type UpdateOptions string

const (
	// Only update existing elements if the new score is less than the current score. Equivalent to
	// "LT" in the Valkey API.
	ScoreLessThanCurrent UpdateOptions = "LT"
	// Only update existing elements if the new score is greater than the current score. Equivalent
	// to "GT" in the Valkey API.
	ScoreGreaterThanCurrent UpdateOptions = "GT"
)

const (
	ChangedKeyword string = "CH"   // Valkey API keyword used to return total number of elements changed
	IncrKeyword    string = "INCR" // Valkey API keyword to make zadd act like ZINCRBY.
)
